File: ESTRUCURAS_DE_DATOS/Cola/pythonEjercicios/TrafficSimulation.py
class CarQueue:
    def __init__(self):
        self.elements = []

    def is_empty(self):
        return len(self.elements) == 0

    def enqueue(self, element):
        self.elements.append(element)

    def dequeue(self):
        if self.elements:
            return self.elements.pop(0)
        return None

    def peek(self):
        if self.elements:
            return self.elements[0]
        return None

    def show_all_cars(self):
        if self.is_empty():
            print("🚫 No hay carros en la cola.")
        else:
            print("🚗 Carros en la cola:")
            for i, car in enumerate(self.elements, start=1):
                print(f"  {i}. {car.brand} - {car.license_plate}")
    


class Car:
    def __init__(self, license_plate, brand):
        self.license_plate = license_plate
        self.brand = brand


class TrafficSimulation:
    def __init__(self, queue):
        self.queue = queue

    def simulate(self):
        count = 0
        while not self.queue.is_empty():
            car = self.queue.dequeue()
            if car:
                count += 1
                print(f"Car #{count}: {car.brand} ({car.license_plate}) was removed from the queue")


# --- Ejecución del programa ---
if __name__ == "__main__":
    queue = CarQueue()
    queue.enqueue(Car('ABC-123', 'Mazda'))
    queue.enqueue(Car('DEF-456', 'Toyota'))
    queue.enqueue(Car('GHI-789', 'Honda'))
    queue.enqueue(Car('JKL-012', 'Ford'))

    queue.show_all_cars()

    print("\n🟢 Iniciando simulación...\n")
    simulation = TrafficSimulation(queue)
    simulation.simulate()

    print("\n🔴 Estado final de la cola:")
    queue.show_all_cars()

File: ESTRUCURAS_DE_DATOS/Cola/typeScriptEjercicios/TrafficSimulation.js
var CarQueue = /** @class */ (function () {
    function CarQueue() {
        this.elements = [];
    }
    CarQueue.prototype.isEmpty = function () {
        return this.elements.length === 0;
    };
    CarQueue.prototype.enqueue = function (element) {
        this.elements.push(element);
    };
    CarQueue.prototype.dequeue = function () {
        return this.elements.shift() || null;
    };
    CarQueue.prototype.peek = function () {
        return this.elements.length > 0 ? this.elements[0] : null;
    };
    return CarQueue;
}());
var Car = /** @class */ (function () {
    function Car(licensePlate, brand) {
        this.licensePlate = licensePlate;
        this.brand = brand;
    }
    return Car;
}());
var TrafficSimulation = /** @class */ (function () {
    function TrafficSimulation(queue) {
        this.queue = queue;
    }
    TrafficSimulation.prototype.simulate = function () {
        var count = 0;
        while (!this.queue.isEmpty()) {
            var car = this.queue.dequeue();
            if (car) {
                count++;
                console.log("Car #".concat(count, ": ").concat(car.brand, " (").concat(car.licensePlate, ") was removed from the queue"));
            }
        }
    };
    return TrafficSimulation;
}());
var queue = new CarQueue();
queue.enqueue(new Car('ABC-123', 'Mazda'));
queue.enqueue(new Car('DEF-456', 'Toyota'));
queue.enqueue(new Car('GHI-789', 'Honda'));
queue.enqueue(new Car('JKL-012', 'Ford'));
var simulation = new TrafficSimulation(queue);
simulation.simulate();
